### ISSUE TITLE
Asynchronous test scheduling

### ISSUE THREAD

**Issue by jugglinmike** [web-platform-tests/wpt#6002]
@jgraham @gsnedders can you please weigh in on this?

Asynchronous tests expressed with the `async_test` function are run in
parallel. If expressed using `promise_test`, however, the tests are run in
series. Demonstration:

```js
'use strict';
var asyncTestGlobal = null;
var promiseTestGlobal = null;

// Fails:
// > assert_equals: expected "first async_test" but got "second async_test"
async_test(function(t) {
    asyncTestGlobal = 'first async_test';
    t.step_timeout(function() {
        assert_equals(asyncTestGlobal, 'first async_test');
        t.done();
      }, 100);
  }, 'first test');

// Passes
async_test(function(t) {
    asyncTestGlobal = 'second async_test';
    t.step_timeout(function() {
        assert_equals(asyncTestGlobal, 'second async_test');
        t.done();
      }, 100);
  }, 'second test');

// Passes
promise_test(function(t) {
    promiseTestGlobal = 'first promise_test';
    return new Promise(function(resolve) {
          t.step_timeout(resolve, 100);
        })
      .then(function() {
          assert_equals(promiseTestGlobal, 'first promise_test');
        });
  }, 'first promise_test');

// Passes
promise_test(function(t) {
    promiseTestGlobal = 'second promise_test';
    return new Promise(function(resolve) {
          t.step_timeout(resolve, 100);
        })
      .then(function() {
          assert_equals(promiseTestGlobal, 'second promise_test');
        });
  }, 'second promise_test');
```

I believe running in series is preferable because it allows authors to more
efficiently work with singletons in the browser. gh-5970 is an example of this,
where the contributor's current options involve either implementing custom test
scheduling logic or expressing tests across multiple files.

(Note that `testharness.js` has no unit tests at the moment, but I've recently
been working to add them in gh-5627. I would strongly discourage changing any
behavior until those tests are in place.)

**Comment by foolip** [comment]
@jugglinmike, what should we do with this issue? This behavior is known, intentional, but possibly not well documented. Can you file issues with suggestions for improvement and close this one?

**Comment by jugglinmike** [comment]
[This is documented via a sentence in the middle of the section on Promise tests](https://github.com/web-platform-tests/wpt/blob/8c024e160ccff17edde16ed3989689d36e6ca085/docs/_writing-tests/testharness-api.md#promise-tests):

> Unlike Asynchronous Tests, Promise Tests don’t start running until after the previous Promise Test finishes.

I'm not sure if we should consider this "well-documented," but I haven't seen much in the way of confusion or complaint (and no one has replied to this issue in the 18 months since I opened it), so maybe it's good enough.

As for improvements, consistency still seems like the answer. I'm not sure it's worth it, though. Making the two APIs consistent would be be quite an undertaking, and if that anecdotal evidence is any indication, it wouldn't solve a serious issue. My preference for serial execution stands, but I'm inclined to let that naturally become the norm as web APIs and test authors warm up to the Promise paradigm.

It's not clear if that sentiment is shared by others, though. If it is, then we could hasten the evolution by deprecating `async_test`. I've opened gh-13530 to discuss that. I expect it to go over like a lead balloon, but it will be helpful to learn why.

That's probably the most productive action to be taken with this issue, so I'll close it now.
